metagraph/src/annotation/annotation_converters.cpp

```diff
@@ -978,5 +980,111 @@ void convert_to_row_annotator(const ColumnCompressed<std::string> &source,
+    std::atomic<uint64_t> forced_terminal_count = 0;
+    std::atomic<uint64_t> depth_terminal_count = 0;
+    std::atomic<uint64_t> boundary_size = 0;
+    std::atomic<uint64_t> visited_nodes = 0;
+    graph.call_sequences(
+        [&](const std::string &, const std::vector<uint64_t> &path) {
+            assert(!path.empty());
+            std::vector<uint64_t> anno_ids(path.size());
+            for (uint32_t i = 0; i < path.size(); ++i) {
+                assert(path[i] <= nnodes);
+                anno_ids[i] = graph::AnnotatedSequenceGraph::graph_to_anno_index(path[i]);
+            }
+            std::vector<Vector<uint64_t>> rows = annotation.get_matrix().get_rows(anno_ids);
+            visited_nodes += path.size();
+            std::sort(rows[0].begin(), rows[0].end());
+            Vector<uint64_t> diffs;
+            for (uint32_t i = 0; i < rows.size() - 1; ++i) {
+                std::sort(rows[i + 1].begin(), rows[i + 1].end());
+
+                // if we reached the max path depth, force a terminal node
+                if (i % max_depth == 0) {
+                    pos[anno_ids[i]] = { diffs.size(), rows[i].size() };
+                    diffs.insert(diffs.end(), rows[i].begin(), rows[i].end());
+                    terminal[anno_ids[i]] = 1;
+                    depth_terminal_count.fetch_add(1, std::memory_order_relaxed);
+                    continue;
+                }
+
+                uint64_t start_idx = diffs.size();
+                std::set_symmetric_difference(rows[i].begin(), rows[i].end(),
+                                              rows[i + 1].begin(), rows[i + 1].end(),
+                                              std::back_inserter(diffs));
+
+                // if we don't gain anything by diffing, make the node terminal
+                if (diffs.size() - start_idx >= rows[i].size()) {
+                    diffs.resize(start_idx);
+                    diffs.insert(diffs.end(), rows[i].begin(), rows[i].end());
+                    terminal[anno_ids[i]] = 1;
+                    forced_terminal_count.fetch_add(1, std::memory_order_relaxed);
+                }
+                pos[anno_ids[i]] = { start_idx, diffs.size() - start_idx };
+            }
+            pos[anno_ids.back()] = { diffs.size(), rows.back().size() };
+            diffs.insert(diffs.end(), rows.back().begin(), rows.back().end());
+            size_t offset;
+#pragma omp critical
+            {
+                offset = node_diffs.size();
+                node_diffs.insert(node_diffs.end(), diffs.begin(), diffs.end());
+            }
+            for (uint32_t i = 0; i < rows.size(); ++i) {
+                pos[anno_ids[i]].first += offset;
+            }
+
+            terminal_count.fetch_add(1, std::memory_order_relaxed);
+            terminal[anno_ids.back()] = 1;
+            // add the last row plus one termination bit for each row
+            boundary_size.fetch_add(diffs.size() + rows.size(), std::memory_order_relaxed);
```

Comment:
If the last node of a contig is a merge, it's not necessary to make it an anchor.

An anchor must be created only if there are no outgoing edges.

This should bring about the same reduction as you got by fetching contigs instead of unitigs (which essentially translates the same idea to forks, and I assume there are about the same number of forks and merges in graphs).

Reply:
Hm, I'm thinking how this can be implemented, and think both forks and merges can be handled the same way here. So you can fetch unitigs from graph and then check if there are outgoing (incoming) edge from the last (first) node.

Why fetch unitigs instead of contigs -- no need to make assumptions about how the contigs are traversed (no need to add that flag `bool select_first_edge`), and also, `pull_unitigs` allows for more efficient parallelization and will have a more scalable implementation in the future.

Reply:
There is an efficient parallel algorithm for this, can discuss in the next meeting.

Basically, we start parallel traversal from random nodes (first, all forks, then all other, as usual), and make every 100-th node an anchor. Also, we mark each traversed node in another bitmap (which indicates the nodes within <=100 steps to the next anchor). And we terminate the traversal once we reach a node which is marked in this extra bitmap.

So we never traverse the same node twice (thanks to the first bitmap of visited nodes), hence the time complexity is the same as for `call_sequences`.
The second bitmap, on the other hand, ensures that there are no paths of size >200 without an anchor.

With some other tricks, we can probably guarantee a lower bound on average too, but this is a bit more tricky.

Also, in fact, we don't need two traversals. Everything can be done in one. So we can implement this in this PR. Basically, this will be a slightly modified version of the current algorithm, with an extra masking bitmap and a few more checks. So creating a variation of the original traversal algorithm as a separate function seems to be the best approach.

Reply:
I prefer having the option to not set maximum path lengths, as the compression is significantly better if there is no max path length (50% by max length 100). If we query not too short sequences, the cost of the path length will be amortized very quickly, so in many applications this makes sense. 

On the other hand, if I am not mistaken, the algorithm you proposed doesn't guarantee that the maximum path length is 200 - in the worst case the maximum path length can be n, where n is the number of edges in the graph (if we keep chaining paths into paths that were already visited).

Reply:
of course, it's a parameter, I used 100 and 200 as an example

Reply:
What I meant was that I don't think this method can provide an upper bound for the length of a path, no matter what the value of the parameter is. In the worst case, you can always have a path of length n (total number of edges), even if you  set the maximum path length to 1.

Reply:
It does, that's why bitmaps, otherwise we would use just one.
If we reach a node marked in the second bitmap while traversing a path, we don't mark the nodes of this path in that second bitmap, because they don't necessarily satisfy the closeness condition (<= 100)

Reply:
> I prefer having the option to not set maximum path lengths, as the compression is significantly better if there is no max path length (50% by max length 100). If we query not too short sequences, the cost of the path length will be amortized very quickly, so in many applications this makes sense.

In this case just set n to a large number, like 1000. But a threshold must be set, otherwise, there are no reasonable guarantees on the query complexity.